airflow/providers/common/sql/hooks/sql.py

```diff
@@ -169,6 +169,33 @@ def get_uri(self) -> str:
         conn.schema = self.__schema or conn.schema
         return conn.get_uri()
 
+    def resolve_rds_cname(self, hostname):
```

Comment:
I don't think it should be a part of Common SQL + Postgres providers.
We have a [discussion previously](https://github.com/apache/airflow/pull/26162#issuecomment-1238159017) that this kind of implementation should exist in particular providers (Amazon in this case)

As well as AWS IAM Auth should be move outside of Postgres/MySQL Hooks.

Maybe this is "Good Time" for thinking about RdsPostgresHook (for Postgres and Aurora) and RdsMySqlHook for (MySQL and Aurora)

cc: @ferruzzi @vincbeck @o-nikolas @potiuk

Reply:
Agree that it is not the best place for that, just wanted to not duplicate the code for PG, MySql, etc. Also, that code path is optional and won't be triggered until we provide `{"resolve_cname": true}` in extra config.

Reply:
There is more to that. Common SQL is a dependency in many providers and we should treat it carefuly. This is AWS-only and I think even now - AWS specific code in the Postgres hook - is wrong. 

I think if we want to add something like that, we should properly use other patterns here - not only inheritance.

I think  we should (and that might be a good experiment on how we can also remove the other "aws-ism" in the generic hooks simply add some "hooks" into the hooks (err better name is needed - maybe a callback) where you would be able to inject AWS-specific behaviours into the generic DB Hooks.

Say in this case we have a generic "host_resolve" callback and when you instantiate Hook you can provide the callback there. This is how it should be done. And I think we should do it now.

Reply:
@potiuk 
Can it be used like this:
```python
 pg_hook  = PostgresHook(..., behaviors=["rds"])
```
and this will enable rds specific logic in the PostgresHook ?

Reply:
Nope. That's not enough. I do not want the AWS code to be in non-AWS provider (even though we have it now in places). 

We have an AWS provider - all code specific to AWS should be there, if/when we split providers out of Airlfow core, I do not want any external service code in "apache-airlfow" package.

The requirement is simple none of AWS-specific code should be in apache/airlfow/providers/common/sql or apache/airflow. All the AWS-specific code should be in apache/airflow/providers/aws/

That's very simple and straighforward expectation. Exactly how this should be done and how the code is to be injected  from one package to the other (via callbacks, hooks, etc. ) - I am open to any proposal.